lib/Dialect/ESI/ESIPasses.cpp

```diff
@@ -84,13 +109,177 @@ void ESIToPhysicalPass::runOnOperation() {
+      WIDTH(Identifier::get("WIDTH", getContext())), declaredStage(nullptr),
+      loc(UnknownLoc::get(getContext())) {
+
+  auto regions = top->getRegions();
+  if (regions.size() == 0) {
+    top->emitError("ESI RTL Builder needs a region to insert RTL.");
+  }
+  auto &region = regions.front();
+  if (!region.empty())
+    setInsertionPoint(&region.front(), region.front().begin());
+}
+
+/// Write an 'ExternModuleOp' to use a hand-coded SystemVerilog module. Said
+/// module implements pipeline stage, adding 1 cycle latency. This particular
+/// implementation is double-buffered and fully pipelines the reverse-flow ready
+/// signal.
+circt::rtl::RTLExternModuleOp ESIRTLBuilder::declareStage() {
+  using namespace circt::rtl;
+
+  if (declaredStage)
+    return declaredStage;
+
+  auto name = StringAttr::get("ESI_PipelineStage", getContext());
+  // Since this module has parameterized widths on the a input and x output,
+  // give the extern declation a None type since nothing else makes sense.
+  // Will be refining this when we decide how to better handle parameterized
+  // types and ops.
+  ArrayRef<ModulePortInfo> ports = {
+      {clk, PortDirection::INPUT, getI1Type(), 0},
+      {rstn, PortDirection::INPUT, getI1Type(), 1},
+      {a, PortDirection::INPUT, getNoneType(), 2},
+      {a_valid, PortDirection::INPUT, getI1Type(), 3},
+      {a_ready, PortDirection::OUTPUT, getI1Type(), 0},
+      {x, PortDirection::OUTPUT, getNoneType(), 1},
+      {x_valid, PortDirection::OUTPUT, getI1Type(), 2},
+      {x_ready, PortDirection::INPUT, getI1Type(), 4}};
+  declaredStage = create<RTLExternModuleOp>(loc, name, ports);
+  return declaredStage;
+}
+
+namespace {
+/// Lower PipelineStage ops to an RTL implementation. Unwrap and re-wrap
+/// appropriately. Another conversion will take care merging the resulting
+/// adjacent wrap/unwrap ops.
+struct PipelineStageLowering : public OpConversionPattern<PipelineStage> {
+public:
+  PipelineStageLowering(ESIRTLBuilder &builder, MLIRContext *ctxt)
+      : OpConversionPattern(ctxt), builder(builder) {}
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(PipelineStage stage, ArrayRef<Value> operands,
+                  ConversionPatternRewriter &rewriter) const final;
+
+private:
+  ESIRTLBuilder &builder;
+};
+} // anonymous namespace
+
+LogicalResult PipelineStageLowering::matchAndRewrite(
+    PipelineStage stage, ArrayRef<Value> stageOperands,
+    ConversionPatternRewriter &rewriter) const {
+  auto loc = stage.getLoc();
+  auto chPort = stage.input().getType().dyn_cast<ChannelPort>();
+  if (!chPort)
+    return failure();
+  auto stageModule = builder.declareStage();
+
+  MutableDictionaryAttr stageAttrs;
+  size_t width = getNumBits(chPort.getInner());
+  stageAttrs.set(builder.WIDTH, rewriter.getUI32IntegerAttr(width));
+
+  // Unwrap the channel. The ready signal is a Value we haven't created yet, so
```

Comment:
Is this the correct way to create back edges? Feels kinda hacky but it works.

Reply:
This is sort of weird but I can't think of a better way to do this off-hand.

Reply:
Could you do something like the following?
```
Value val;
...
[use `val` as needed]
...
val = [code to create value]
```
I'm guessing passing around a null Value might not work and this temp constant was needed?

Reply:
> I'm guessing passing around a null Value might not work and this temp constant was needed?

Spot on.

Reply:
https://llvm.discourse.group/t/canonical-way-to-create-back-edges/2415/2